### PR TITLE
feat: cypress requests mocking

### PR DESCRIPTION
**Pull request by vividwood**
## Description

Added autorecording and stubbing to improve testing time

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes https://github.com/shapeshift/web/issues/1554

## Screenshots (if applicable)

https://user-images.githubusercontent.com/101632255/165683707-9ab210e2-2757-498f-bb24-95baf4ea535d.mp4

### CODE REVIEW

File: cypress/plugins/index.ts
Comment (0xApotheosis) [inline cypress/plugins/index.ts:57]: Add comments above each of these lines explaining what they are doing.

File: cypress/plugins/index.ts
Comment (0xApotheosis) [inline cypress/plugins/index.ts:59]: Add a comment on what this is doing.

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Thanks for the initial PR @vividwood, we are off to a good start. I'm going to dig into the implementation more, but for now a few high level notes:

1. At the moment we are only mocking the request in `dashboard_spec`. We'll need to mock them for all tests. I like that there is only one `.har` file - I'd expect to see a `All Integration Specs.har` file instead of the `dashboard_spec.har` file we have currently.

2. The requests in the `beforeAll` hooks don't appear to be mocked, and take a relatively long time (see screenshot below). I expect we can improve runtime significantly by getting these mocked.

Screen Shot 2022-05-02 at 4 59 01 pm

3. I don't actually see much improvement in the test runtime (it actually got slower when running all tests), which might be something to do with additional setup overhead to load in the HAR file. It will be interesting to see the difference once all requests are mocked, not just those in `dashboard.spec`.

4. We need to have a workflow on how these HAR files get updated. Can we use a flag we can pass in that when enabled will update the HAR and allow us to commit it to version control? Once we've got this process it'll need to be documented in the README.

**Comment by vividwood** [comment]
> Thanks for the initial PR @vividwood, we are off to a good start. I'm going to dig into the implementation more, but for now a few high level notes:
> 
>     1. At the moment we are only mocking the request in `dashboard_spec`. We'll need to mock them for all tests. I like that there is only one `.har` file - I'd expect to see a `All Integration Specs.har` file instead of the `dashboard_spec.har` file we have currently.
> 
>     2. The requests in the `beforeAll` hooks don't appear to be mocked, and take a relatively long time (see screenshot below). I expect we can improve runtime significantly by getting these mocked.
> 
> 
> Screen Shot 2022-05-02 at 4 59 01 pm
> 
>     3. I don't actually see much improvement in the test runtime (it actually got slower when running all tests), which might be something to do with additional setup overhead to load in the HAR file. It will be interesting to see the difference once all requests are mocked, not just those in `dashboard.spec`.
> 
>     4. We need to have a workflow on how these HAR files get updated. Can we use a flag we can pass in that when enabled will update the HAR and allow us to commit it to version control? Once we've got this process it'll need to be documented in the README.

2. Yes, I left it because I thought we don't need to mock login details, it starts mocking everything starting from `beforeEach`. It's easy to fix. 

3. I think you don't see it because when starting all tests, it tries to read from `All Integration Specs.har` which don't exists right now. Aside of that, there is a big improvement in tests except first one (because of page loading I guess), like `support trades` went from 16s to 4s. 

4. I can add some `force` flag which will overwrite entire `har` file.

**Comment by 0xApotheosis** [comment]
On 2, yes please - mock everything.

On 3, indeed `All Integration Specs.har` doesn't exist - we need to add it as part of this PR (and remove the dashboard-only one). All tests should be mocked when the whole test suite is run.

**Comment by 0xdef1cafe** [comment]
@vividwood are you codinglittlecat/visitskyworld?

**Comment by vividwood** [comment]
> @vividwood are you codinglittlecat/visitskyworld?

Nope

**Comment by 0xdef1cafe** [comment]
@vividwood can you please provide an ETA for when you'll finish this PR?

**Comment by vividwood** [comment]
> @vividwood can you please provide an ETA for when you'll finish this PR?

It is. I'm waiting for a review from @0xApotheosis

**Comment by 0xApotheosis** [comment]
> > @vividwood can you please provide an ETA for when you'll finish this PR?
> 
> It is. I'm waiting for a review from @0xApotheosis

Yep, @0xdef1cafe this is blocked on me doing an extensive review.

@vividwood in the interim I'm noticing there isn't yet a README update on the process to update the mocks.

**Comment by vividwood** [comment]
> > > @vividwood can you please provide an ETA for when you'll finish this PR?
> > 
> > 
> > It is. I'm waiting for a review from @0xApotheosis
> 
> Yep, @0xdef1cafe this is blocked on me doing an extensive review.
> 
> @vividwood in the interim I'm noticing there isn't yet a README update on the process to update the mocks.

Added

File: cypress.json
Comment (0xApotheosis) [inline cypress.json:14]: Add a paragraph in the README on what each of these settings do.

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
All looking good, and working well @visitskyworld! A nice implementation.

As discussed, a few small improvements to make:

1. Move some of the complicated conditional logic into constants with descriptive names
2. Update the README with information on each of the new autorecord options, e.g. `cleanMocks`.

File: package.json
Comment (vividwood) [reply package.json:112]: > This doesn't work on my machine (macOS), should we use `chrome` instead of `chromium` to match CI?
> 
> ```
> yarn run v1.22.18
> $ yarn cypress run --browser chromium --headless
> $ /web/node_modules/.bin/cypress run --browser chromium --headless
> Can't run because you've entered an invalid browser name.
> 
> Browser: chromium was not found on your system or is not supported by Cypress.
> 
> Cypress supports the following browsers:
>  - electron
>  - chrome
>  - chromium
>  - chrome:canary
>  - edge
>  - firefox
> 
> You can also use a custom browser: https://on.cypress.io/customize-browsers
> 
> Available browsers found on your system are:
>  - chrome
>  - electron
>  - Brave
> ```

Actually, this for for `har` thing, it crashes without it. The reason I kept it was just to make it more explicit. So we can revert it back or just keep `--headless` at least.

File: cypress/support/commands.ts
Comment (vividwood) [reply cypress/support/commands.ts:84]: > It's not clear to me why we need this. You mentioned something about "timestamp-based requests", could you please add a comment here explaining the rationale?

Some requests use timestamps in query, so every time you run test they are different. You can't cache them unless you set fixed timestamp

**Comment by vividwood** [comment]
> All looking good, and working well @visitskyworld! A nice implementation.
> 
> As discussed, a few small improvements to make:
> 
>     1. Move some of the complicated conditional logic into constants with descriptive names
> 
>     2. Update the README with information on each of the new autorecord options, e.g. `cleanMocks`.

My nickname is @vividwood not @visitskyworld :)

**Comment by 0xApotheosis** [comment]
> My nickname is @vividwood not @visitskyworld :)

Whoops, right you are @vividwood! Pressed the first name that came up when typing 'v', also the same name as the last bounty hunter that worked on this - apologies 😅

File: cypress/plugins/autorecord/index.ts
Comment (0xApotheosis) [inline cypress/plugins/autorecord/index.ts:52]:
```suggestion
  const cleanMockData: Record<string, unknown> = {}
```

File: cypress/plugins/autorecord/index.ts
Comment (0xApotheosis) [inline cypress/plugins/autorecord/index.ts:54]:
```suggestion
  let routesByTestId: Record<string, unknown> = {}
```

File: cypress/plugins/autorecord/index.ts
Comment (0xApotheosis) [inline cypress/plugins/autorecord/index.ts:56]:
```suggestion
  let routes: Record<string, unknown>[] = []
```

Comment (0xApotheosis) [reply cypress/plugins/autorecord/index.ts:56]: Or even better, can we properly type this?

Even value has `{ url, method, status, data, body, headers }`, yeah?

File: cypress/support/commands.ts
Comment (0xApotheosis) [inline cypress/support/commands.ts:85]: Thanks for the explanation!

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Thanks for adding the additional descriptions!

A few more comments on types.

File: cypress/plugins/autorecord/index.ts
Comment (0xApotheosis) [inline cypress/plugins/autorecord/index.ts:109]: We can keep this pure by making it an IIFE, `const data = (() => {...`

Comment (vividwood) [reply cypress/plugins/autorecord/index.ts:109]: Good point

**Review by 0xApotheosis — APPROVED** [review]
It's a go from me, @vividwood 🍬 

We've got most requests successfully mocked in an automated way, and a nice mechanism for updating and regenerating the mocks where needed.

The serialised responses keep our diff nice and tight, which is a win.

Thanks for your help on this one!

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**
